File: ambitapp-v2/src/services/weatherservice.cpp
#include "weatherservice.h"

#include <QJsonArray>
#include <QJsonDocument>
#include <QJsonObject>
#include <QNetworkReply>
#include <QUrlQuery>

WeatherService::WeatherService(QObject *parent) : QObject(parent)
{
    // Background auto-refresh, real request 2026-08-07: every 10 minutes, independent of
    // whoever else calls refresh() (HomePage.qml on load, Settings after a location change).
    m_refreshTimer.setInterval(10 * 60 * 1000);
    connect(&m_refreshTimer, &QTimer::timeout, this, &WeatherService::refresh);
    m_refreshTimer.start();
}

void WeatherService::setLoading(bool value)
{
    if (m_loading == value)
        return;
    m_loading = value;
    emit loadingChanged();
}

void WeatherService::setAvailable(bool value)
{
    if (m_available == value)
        return;
    m_available = value;
    emit availableChanged();
}

void WeatherService::markFetchedOnce()
{
    if (m_hasFetchedOnce)
        return;
    m_hasFetchedOnce = true;
    emit hasFetchedOnceChanged();
}

void WeatherService::setLatitude(double value)
{
    if (qFuzzyCompare(m_latitude, value))
        return;
    m_latitude = value;
    emit locationChanged();
}

void WeatherService::setLongitude(double value)
{
    if (qFuzzyCompare(m_longitude, value))
        return;
    m_longitude = value;
    emit locationChanged();
}

void WeatherService::refresh()
{
    setLoading(true);

    QUrl url(QStringLiteral("https://api.open-meteo.com/v1/forecast"));
    QUrlQuery query;
    query.addQueryItem(QStringLiteral("latitude"), QString::number(m_latitude));
    query.addQueryItem(QStringLiteral("longitude"), QString::number(m_longitude));
    query.addQueryItem(QStringLiteral("current"),
                        QStringLiteral("temperature_2m,weather_code,wind_speed_10m"));
    query.addQueryItem(QStringLiteral("daily"),
                        QStringLiteral("temperature_2m_max,temperature_2m_min,weather_code"));
    query.addQueryItem(QStringLiteral("timezone"), QStringLiteral("auto"));
    query.addQueryItem(QStringLiteral("forecast_days"), QStringLiteral("3"));
    url.setQuery(query);

    QNetworkReply *reply = m_network.get(QNetworkRequest(url));
    connect(reply, &QNetworkReply::finished, this, [this, reply] {
        reply->deleteLater();
        setLoading(false);

        // Any failure here, network or malformed response, just leaves `available` false;
        // WeatherCard.qml shows a plain "you're offline" message rather than any raw error.
        if (reply->error() != QNetworkReply::NoError) {
            setAvailable(false);
            markFetchedOnce();
            return;
        }

        const auto doc = QJsonDocument::fromJson(reply->readAll());
        const auto root = doc.object();
        const auto current = root.value(QStringLiteral("current")).toObject();
        const auto daily = root.value(QStringLiteral("daily")).toObject();
        const auto dailyTime = daily.value(QStringLiteral("time")).toArray();
        const auto dailyHigh = daily.value(QStringLiteral("temperature_2m_max")).toArray();
        const auto dailyLow = daily.value(QStringLiteral("temperature_2m_min")).toArray();
        const auto dailyCode = daily.value(QStringLiteral("weather_code")).toArray();

        if (current.isEmpty() || dailyTime.isEmpty()) {
            setAvailable(false);
            markFetchedOnce();
            return;
        }

        m_currentTemperature = current.value(QStringLiteral("temperature_2m")).toDouble();
        m_currentWeatherCode = current.value(QStringLiteral("weather_code")).toInt();
        m_windSpeed = current.value(QStringLiteral("wind_speed_10m")).toDouble();
        m_todayHigh = dailyHigh.isEmpty() ? 0 : dailyHigh.at(0).toDouble();
        m_todayLow = dailyLow.isEmpty() ? 0 : dailyLow.at(0).toDouble();

        m_forecast.clear();
        for (int i = 0; i < dailyTime.size(); ++i) {
            QVariantMap day;
            day[QStringLiteral("date")] = dailyTime.at(i).toString();
            day[QStringLiteral("high")] = i < dailyHigh.size() ? dailyHigh.at(i).toDouble() : 0;
            day[QStringLiteral("low")] = i < dailyLow.size() ? dailyLow.at(i).toDouble() : 0;
            day[QStringLiteral("code")] = i < dailyCode.size() ? dailyCode.at(i).toInt() : -1;
            m_forecast.append(day);
        }

        setAvailable(true);
        markFetchedOnce();
        emit dataChanged();
    });

    fetchPlaceName();
}

void WeatherService::fetchPlaceName()
{
    QUrl url(QStringLiteral("https://nominatim.openstreetmap.org/reverse"));
    QUrlQuery query;
    query.addQueryItem(QStringLiteral("lat"), QString::number(m_latitude));
    query.addQueryItem(QStringLiteral("lon"), QString::number(m_longitude));
    query.addQueryItem(QStringLiteral("format"), QStringLiteral("json"));
    query.addQueryItem(QStringLiteral("zoom"), QStringLiteral("10"));  // city/town level
    url.setQuery(query);

    QNetworkRequest request(url);
    // Nominatim's usage policy requires a real identifying User-Agent, not the Qt default -
    // see https://operations.osmfoundation.org/policies/nominatim/.
    request.setHeader(QNetworkRequest::UserAgentHeader, QStringLiteral("AmbitApp/2.0"));

    QNetworkReply *reply = m_network.get(request);
    connect(reply, &QNetworkReply::finished, this, [this, reply] {
        reply->deleteLater();
        if (reply->error() != QNetworkReply::NoError) {
            return;  // same "just don't show it" rule as weather data itself
        }
        const auto doc = QJsonDocument::fromJson(reply->readAll());
        const auto address = doc.object().value(QStringLiteral("address")).toObject();
        // Nominatim's address breakdown has no single "the name" field - city/town/village
        // is the real hierarchy for how populated places are actually classified there, in
        // that fallback order, matching what a weather card should show regardless of
        // whether the coordinate lands in a big city or a small village.
        const QString name = [&]() {
            for (const auto &key : {"city", "town", "village", "municipality", "county"}) {
                const auto value = address.value(QLatin1String(key)).toString();
                if (!value.isEmpty())
                    return value;
            }
            return QString();
        }();
        if (m_placeName != name) {
            m_placeName = name;
            emit placeNameChanged();
        }
    });
}

void WeatherService::detectLocationFromIp()
{
    // ip-api.com's free tier: no key, plain HTTP only (no HTTPS on the free tier), a real,
    // commonly-used service. Approximate (city-level, from the IP's registered location,
    // not true GPS) - the honest limit of anything IP-based, stated here rather than implied
    // to be more precise than it is.
    QNetworkReply *reply = m_network.get(
        QNetworkRequest(QUrl(QStringLiteral("http://ip-api.com/json/"))));
    connect(reply, &QNetworkReply::finished, this, [this, reply] {
        reply->deleteLater();
        // This is the real startup path now (HomePage.qml calls this instead of refresh()
        // directly), so a failed IP lookup still has to end in a real refresh() call with
        // whatever coordinates are already set (the static central-Europe fallback, or
        // wherever the user last pointed it via Settings) - otherwise hasFetchedOnce would
        // never become true on a genuinely offline launch, and WeatherCard.qml would stay
        // hidden forever instead of showing its real offline message.
        if (reply->error() != QNetworkReply::NoError) {
            refresh();
            return;
        }
        const auto obj = QJsonDocument::fromJson(reply->readAll()).object();
        if (obj.value(QStringLiteral("status")).toString() != QStringLiteral("success")) {
            refresh();
            return;
        }
        setLatitude(obj.value(QStringLiteral("lat")).toDouble());
        setLongitude(obj.value(QStringLiteral("lon")).toDouble());
        refresh();
    });
}

File: android/src/services/WeatherService.ts
import { useEffect, useRef, useState } from 'react';

// v3.0 UI port (2026-08-09, "replicate the desktop version feature wise") - ports
// desktop/src/services/weatherservice.cpp + qml/viewmodels/WeatherViewModel.qml byte-for-byte
// in intent: same Open-Meteo endpoint/params, same Nominatim reverse-geocode, same
// ip-api.com IP-based location fallback, same WMO weather-code -> icon/label mapping, same
// 10-minute auto-refresh, same "hide until hasFetchedOnce, then show a friendly offline
// message rather than an error" rule. No API key anywhere, matching AMBITAPP_SPEC.md's own
// "Open-Meteo (no API key)" choice - see weatherservice.h's own header comment.

export interface ForecastDay {
  date: string;
  high: number;
  low: number;
  code: number;
}

export interface WeatherData {
  currentTemperature: number;
  currentWeatherCode: number;
  windSpeed: number;
  todayHigh: number;
  todayLow: number;
  forecast: ForecastDay[];
}

const REFRESH_INTERVAL_MS = 10 * 60 * 1000;
// Same real "no location configured yet" placeholder as weatherservice.h's own
// m_latitude/m_longitude defaults - roughly central Europe, not 0,0 (the Gulf of Guinea).
const DEFAULT_LAT = 48.85;
const DEFAULT_LON = 2.35;

export async function fetchWeather(latitude: number, longitude: number): Promise<WeatherData | null> {
  // URLSearchParams + string concat, not `new URL()` - matches ApiStrava.ts's
  // own established pattern in this codebase (URL's own constructor isn't used anywhere
  // else here).
  const params = new URLSearchParams({
    latitude: String(latitude),
    longitude: String(longitude),
    current: 'temperature_2m,weather_code,wind_speed_10m',
    daily: 'temperature_2m_max,temperature_2m_min,weather_code',
    timezone: 'auto',
    forecast_days: '3',
  });
  const res = await fetch(`https://api.open-meteo.com/v1/forecast?${params.toString()}`);
  if (!res.ok) return null;
  const body = await res.json();
  const current = body?.current;
  const daily = body?.daily;
  if (!current || !daily?.time?.length) return null;

  const forecast: ForecastDay[] = daily.time.map((date: string, i: number) => ({
    date,
    high: daily.temperature_2m_max?.[i] ?? 0,
    low: daily.temperature_2m_min?.[i] ?? 0,
    code: daily.weather_code?.[i] ?? -1,
  }));

  return {
    currentTemperature: current.temperature_2m ?? 0,
    currentWeatherCode: current.weather_code ?? -1,
    windSpeed: current.wind_speed_10m ?? 0,
    todayHigh: forecast[0]?.high ?? 0,
    todayLow: forecast[0]?.low ?? 0,
    forecast,
  };
}

export async function fetchPlaceName(latitude: number, longitude: number): Promise<string> {
  const params = new URLSearchParams({
    lat: String(latitude),
    lon: String(longitude),
    format: 'json',
    zoom: '10',
  });

  try {
    // Nominatim's usage policy requires a real identifying User-Agent - same real header
    // weatherservice.cpp's own fetchPlaceName() sends.
    const res = await fetch(`https://nominatim.openstreetmap.org/reverse?${params.toString()}`, {
      headers: { 'User-Agent': 'Sommet/2.0' },
    });
    if (!res.ok) return '';
    const body = await res.json();
    const address = body?.address ?? {};
    for (const key of ['city', 'town', 'village', 'municipality', 'county']) {
      if (address[key]) return address[key];
    }
    return '';
  } catch {
    return '';
  }
}

export async function detectLocationFromIp(): Promise<{ latitude: number; longitude: number } | null> {
  try {
    // ip-api.com's free tier: no key, HTTP only (no HTTPS) - same real, commonly-used
    // service weatherservice.cpp's own detectLocationFromIp() calls. Approximate
    // (city-level from the IP's registered location, not true GPS).
    const res = await fetch('http://ip-api.com/json/');
    if (!res.ok) return null;
    const body = await res.json();
    if (body?.status !== 'success') return null;
    return { latitude: body.lat, longitude: body.lon };
  } catch {
    return null;
  }
}

// Real WMO weather-interpretation codes (open-meteo.com/en/docs, "WMO Weather interpretation
// codes") - same ranges as WeatherViewModel.qml's own iconFor()/labelFor(), emoji instead of
// this app's own Material-Symbols-subset Icon.tsx (which has no weather glyphs at all - see
// assets/fonts/NOTICE.md on why that font is deliberately small/subsetted; adding 7 new
// glyphs to it isn't worth it for a widget emoji already render natively).
export function weatherEmoji(code: number): string {
  if (code === 0) return '☀️';
  if (code === 1 || code === 2) return '🌤️';
  if (code === 3) return '☁️';
  if (code === 45 || code === 48) return '🌫️';
  if (code >= 51 && code <= 67) return '🌧️';
  if (code >= 80 && code <= 82) return '🌧️';
  if (code === 71 || code === 73 || code === 75 || code === 77) return '❄️';
  if (code === 85 || code === 86) return '❄️';
  if (code === 95 || code === 96 || code === 99) return '⛈️';
  return '🌤️';
}

interface WeatherState {
  loading: boolean;
  available: boolean;
  hasFetchedOnce: boolean;
  placeName: string;
  data: WeatherData | null;
}

// Component-scoped (not a module-level singleton like WeatherService.cpp's own QML
// singleton) - Home is the only real consumer right now, so a hook owning its own state +
// refresh timer for as long as Home is mounted is the simplest thing that's still correct;
// a shared singleton would only matter once a second screen needs the same live data.
export function useWeather() {
  const [state, setState] = useState<WeatherState>({
    loading: false, available: false, hasFetchedOnce: false, placeName: '', data: null,
  });
  const locationRef = useRef({ latitude: DEFAULT_LAT, longitude: DEFAULT_LON });

  async function refresh() {
    setState(s => ({ ...s, loading: true }));
    const { latitude, longitude } = locationRef.current;
    const [data, placeName] = await Promise.all([
      fetchWeather(latitude, longitude),
      fetchPlaceName(latitude, longitude),
    ]);
    setState({
      loading: false,
      available: data !== null,
      hasFetchedOnce: true,
      placeName,
      data,
    });
  }

  async function detectAndRefresh() {
    const loc = await detectLocationFromIp();
    if (loc) locationRef.current = loc;
    await refresh();
  }

  useEffect(() => {
    detectAndRefresh();
    const iv = setInterval(refresh, REFRESH_INTERVAL_MS);
    return () => clearInterval(iv);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return { ...state, refresh };
}
